### PR TITLE
- Adding training intermediate verification.

### LINKED ISSUE #1667: Enable golden verification for training
Currently we have golden intermediate verification for forward pass. 
For successful debugging of training graph (forward, backward, and optimizer) we need following golden comparisons:

intermediate gradients (gradients with respect to outputs of intermediate operations),
parameter gradients,
input (activations) gradients (if inputs require grad),
parameter updates (if optimizer is on device)

### PR DESCRIPTION
**Pull request by dgolubovicTT**
### Ticket
Closes #1667 

### What's changed
- Training verification includes comparison of:
		intermediate gradients (gradients with respect to outputs of intermediate operations),
		parameter gradients,
		input (activations) gradients (if inputs require grad),
		parameter updates (if optimizer is on device)
- Adding a small end to end test with verification included.
- Cleanup redundant code. Remove redundant calls to calculate_grads and _run_pytorch_backward. Calling backward once for output tensor is enough to have golden gradients for comparison

Training verification can be performed in compile stages after autograd pass (since before we don't have backward graph). Graph is evaled op by op (forward and backward) and corresponding gradients are compared to the result of pytorch backward.

Few implementation details not to miss:
- Previously, compile context attributes `context.outputs `and `context.parameter_dict` were not from the same computational graph. Namely, context.outputs was output of ForgeModule forward, while context.parameter_dict was from PytorchModule. So essentially, we took PytorchModule extracted parameters, then translated to ForgeModule and took outputs from forward. This is why outputs.backward() didn't produce change in parameter gradients… 
Solution: use ForgeModule to extract parameter_dict

- From now, we pass optimizer to do_verify function. Optimizer parameters are inputs to the graph, so we use them to calculate updates of trainable parameters.

### CODE REVIEW

**Comment by dgolubovicTT** [comment]
Here is how it looks in log:
intermediate_training_verification

File: forge/forge/compile.py [hunk 1]
```diff
@@ -302,6 +302,12 @@ def forge_compile_from_context(context: CompileContext) -> CompiledModel:
 
         # Execute the current stage.
         next_stage = stage_to_func[current_stage](context)
+        in_training = context.training and current_stage.value >= CompileDepth.AUTOGRAD.value
+
+        #  Set optimizer parameters from module, because we need optimizer parameters in intermediate verification.
+        if in_training and context.optimizer_on_device():
+            module_params = context.modules[0].get_parameters()
+            context.optimizer.set_parameters_to_optimize(module_params)
```
Comment (pilkicTT) [inline forge/forge/compile.py:310]: this should be done once in `run_autograd_pass`, not here

File: forge/forge/compile.py [hunk 2]
```diff
@@ -316,7 +322,6 @@ def forge_compile_from_context(context: CompileContext) -> CompiledModel:
         if (
             verify_cfg.verify_all or (verify_cfg.verify_last and should_early_stop_compilation) or should_verify
         ) and can_verify:
-            in_training = context.compiler_cfg.enable_training and current_stage.value >= CompileDepth.AUTOGRAD.value
             do_verify(
                 current_stage.name.lower(),
                 in_training,
```
Comment (pilkicTT) [inline forge/forge/compile.py:327]: it feels like the `do_verify()` should figure out this by itself... it has a graph, so it can check if we have bwd subgraph in there, etc...

File: forge/forge/verify/verify.py [hunk 3]
```diff
@@ -37,33 +37,21 @@
...
+def _run_pytorch_backward(outputs, losses):
     retain_graph = True
     for i, o in enumerate(outputs):
         if o.requires_grad:
-            if device.loss_module is None:
-                loss = narrow_forge_tensor_to_pytorch(losses[i], o.value().shape)
-                o.value().backward(loss, retain_graph=retain_graph)
-            else:
-                o.value().backward(retain_graph=True)  # this is loss
+            loss = narrow_forge_tensor_to_pytorch(losses[i], o.value().shape)
```
[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/forge/verify/verify.py:53]: we can do this in a separate PR, of course...

**Review by pilkicTT — CHANGES_REQUESTED** [review]

File: forge/forge/verify/verify.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/forge/verify/verify.py:53]: Yeah, I have a list of things related to pybuda that should be cleaned as a follow up here. As we talked, I opted to do it in a separate PR. Will create an issue

File: forge/forge/compile.py [hunk 4]
```diff
@@ -168,11 +168,11 @@ def calculate_grads(outputs: Tuple[Tensor, ...], intermediate_golden_tensors: Di
 
     if not losses or run_backward:
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/forge/compile.py:169]: Good catch, we don't need it now :)

File: forge/forge/compile.py (hunk 1 above)
Comment (dgolubovicTT) [reply forge/forge/compile.py:310]: True

File: forge/csrc/graph_lib/python_bindings.cpp [hunk 5]
```diff
@@ -1389,60 +1388,59 @@ std::unordered_map<std::string, py::object> get_graph_input_mapping(
...
+            // E.g. for 'l1.weight' trainable parameter, the optimizer parameters could be learning rate and momentum.
+            // If that was the only trainable parameter in our example, the optimizer_params dictionary would look like
+            // this:
+            // {
+            // ('l1.weight', 'lr'): 0.01,
+            // ('l1.weight', 'momentum'): 0.9,
+            // }
+            // So key to this dictionary is a tuple of parameter name and optimizer parameter name.
+
+            if (optimizer_params.is_none())
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/csrc/graph_lib/python_bindings.cpp:1416]: You are right, inside this if `input->is_optimizer_parameter()` `optimizer_params.is_none()` is not a valid state. I can just add an assert

File: forge/forge/compile.py (hunk 2 above)
Comment (dgolubovicTT) [reply forge/forge/compile.py:327]: There is mapping for each node Id to SubgraphId `node_id_to_subgraph_id_`. However, I don't see which subgraph_id is Backward. 
We have 
```
enum class SubgraphType
{
    Forward,
    Backward,
    Loss,
    Optimizer,
};
```
but seems we are not using it. I figured I could use `nodes_by_subgraph` if I knew the Id of Backward graph, but I don't. Can you give some insight?

Comment (pilkicTT) [reply forge/forge/compile.py:317]: i think you should be able to use `graph->contains_bwd_nodes()` - if it does, then it's in training...

File: forge/forge/verify/verify.py [hunk 6]
```diff
@@ -37,33 +37,21 @@
...
+            o.value().backward(loss, retain_graph=retain_graph)
```
[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/forge/verify/verify.py:51]: hm, i am still not sure, because whenever you call `backward()` on a tensor, it should create its own graph, as far as i understand...

maybe we should test this scenario?

[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/forge/verify/verify.py:51]: Here is a pytorch example:

```
import torch

x = torch.tensor(2.0, requires_grad=True)
y = x * x

# First backward pass
y.backward()  # OK

# Second backward pass on same output
try:
    y.backward() 
except RuntimeError as e:
    print("Expected error:", e)
```

Here is error when trying to call .backward second time:

`Expected error: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.`

File: forge/forge/compile.py (hunk 2 above)
Comment (dgolubovicTT) [reply forge/forge/compile.py]: Done, tnx

File: forge/forge/verify/verify.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/forge/verify/verify.py:51]: but this is calling `backward()` on the same tensor... you are mentioning multiple outputs, i.e. different tensors, so i was wondering if it is necessary in this case?

File: forge/test/mlir/test_golden_verification.py [hunk 7]
```diff
@@ -74,3 +77,45 @@ def forward(self, x):
...
+        torch.rand(shape2, requires_grad=train),
+    ]
+
+    framework_model = Matmul()
+    framework_model.eval() if not train else framework_model.train()
+    verify_cfg = DepricatedVerifyConfig()
+    # Use verify_all to verify all stages of compilation.
+    # verify_cfg.verify_all = True
+    # Use this stages_for_intermediate_verification to specify stage in which you want to verify.
+    verify_cfg.stages_for_intermediate_verification = {CompileDepth.AUTOGRAD}
```
Comment (pilkicTT) [inline forge/test/mlir/test_golden_verification.py:112]: can we have a test which will test golden verification on each stage individually and also when `verify_all=True`?

**Review by pilkicTT — APPROVED** [review]

File: forge/test/mlir/test_golden_verification.py (hunk 7 above)
Comment (dgolubovicTT) [reply forge/test/mlir/test_golden_verification.py]: done
